File: java/src/main/java/nl/craftsmen/workshops/reactivemeetup/util/RailwayStreams.java
package nl.craftsmen.workshops.reactivemeetup.util;

import nl.craftsmen.workshops.reactivemeetup.domain.railway.RailwayStation;
import nl.craftsmen.workshops.reactivemeetup.domain.railway.StationaryTrainSimulation;
import nl.craftsmen.workshops.reactivemeetup.domain.railway.CompositeTrainSimulation;
import nl.craftsmen.workshops.reactivemeetup.domain.railway.GateCheckEvent;
import nl.craftsmen.workshops.reactivemeetup.domain.railway.TrainJourneySimulation;
import nl.craftsmen.workshops.reactivemeetup.domain.railway.TrainSimulationParameters;
import nl.craftsmen.workshops.reactivemeetup.domain.railway.TrainMetrics;
import nl.craftsmen.workshops.reactivemeetup.domain.railway.TrainSimulation;
import nl.craftsmen.workshops.reactivemeetup.domain.railway.TravelCostMatrix;

import rx.Observable;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.concurrent.TimeUnit;

import static nl.craftsmen.workshops.reactivemeetup.util.Utils.sample;

public class RailwayStreams {

	private static final DateFormat DATE_FORMAT = new SimpleDateFormat("HH:mm:ss.SSS dd-MM-YYYY");

	private static Observable<GateCheckEvent> gateCheckEvent$;

	public static Observable<GateCheckEvent> gateCheckEvent$() {

		if (gateCheckEvent$ == null) {
			gateCheckEvent$ = Observable.merge(Arrays.asList(
				singleGateCheckEvent$(true, 233),
				singleGateCheckEvent$(true, 978),
				singleGateCheckEvent$(false, 1313),
				singleGateCheckEvent$(true, 2105),
				singleGateCheckEvent$(false, 3643),
				singleGateCheckEvent$(false, 4411),
				singleGateCheckEvent$(true, 5556),
				singleGateCheckEvent$(false, 8123),
				singleGateCheckEvent$(false, 9722),
				singleGateCheckEvent$(true, 10880)
			));
		}

		return gateCheckEvent$;
	}

	private static Observable<GateCheckEvent> singleGateCheckEvent$(boolean isCheckin, long delay) {
		return Observable.from(Arrays.asList(
			new GateCheckEvent(isCheckin, System.currentTimeMillis() + delay, RailwayStation.AMR))
		).delay(delay, TimeUnit.MILLISECONDS);
	}

	public static Observable<GateCheckEvent> personalCheckinsCheckouts$() {
		return sample(Observable.from(Arrays.asList(
			new GateCheckEvent(true, parseDate("08:04:11.345 16-12-2016"), RailwayStation.UTR),
			new GateCheckEvent(false, parseDate("08:41:03.409 16-12-2016"), RailwayStation.AMS),
			new GateCheckEvent(true, parseDate("17:44:56.122 16-12-2016"), RailwayStation.AMS),
			new GateCheckEvent(false, parseDate("18:49:04.123 16-12-2016"), RailwayStation.DH),
			new GateCheckEvent(true, parseDate("22:15:44.616 16-12-2016"), RailwayStation.DH),
			new GateCheckEvent(true, parseDate("08:03:54.883 17-12-2016"), RailwayStation.UTR),
			new GateCheckEvent(false, parseDate("08:39:21.512 17-12-2016"), RailwayStation.AMS))),
		500);
	}

	public static TravelCostMatrix travelCostMatrix() {
		return TravelCostMatrix.builder()
			.define(RailwayStation.UTR, RailwayStation.AMS, 7.50)
			.define(RailwayStation.UTR, RailwayStation.DH, 11.00)
			.define(RailwayStation.DH, RailwayStation.AMS, 11.50)
			.build();
	}

	public static Observable<TrainMetrics> trainMetrics$() {

		TrainSimulationParameters simulationParameters = new TrainSimulationParameters()
			.setTickFrequency(100)
			.setMaxVelocity(140 / 3.6)
			.setAcceleration(2 / 3.6)
			.setTrainId("1042")
			.setTimeDilation(40.0);

		TrainSimulation simulation = new CompositeTrainSimulation(
			new StationaryTrainSimulation(RailwayStation.AMR.getLocation(), 60.0),
			new TrainJourneySimulation(RailwayStation.AMR.getLocation(), RailwayStation.UTR.getLocation()),
			new StationaryTrainSimulation(RailwayStation.UTR.getLocation(), 60.0)
		);
		
		return simulation.trainMetrics$(simulationParameters, System.currentTimeMillis());
	}

	private static long parseDate(String dateString) {
		try {
			return DATE_FORMAT.parse(dateString).getTime();
		} catch (ParseException e) {
			throw new RuntimeException(e);
		}
	}

}

File: javascript/src/util/railway-streams.js
var Rx = require('rxjs/Rx');
var sample = require('./utils').sample;
var TravelCostMatrix = require('../domain/railway/travel-cost-matrix');
var GateCheckEvent = require('../domain/railway/gate-check-event');
var TrainSimulationParameters = require('../domain/railway/train-simulation-parameters');
var StationaryTrainSimulation = require('../domain/railway/stationary-train-simulation');
var TrainJourneySimulation = require('../domain/railway/train-journey-simulation');
var CompositeTrainSimulation = require('../domain/railway/composite-train-simulation');
var RailwayStation = require('../domain/railway/railway-station');

let gce;

exports.gateCheckEvents$ = gateCheckEvent$();

exports.personalCheckinsCheckouts$ = sample(Rx.Observable.of(
    new GateCheckEvent(true, new Date(2016, 12, 16, 8, 4, 11, 345), 'UTR'),
    new GateCheckEvent(false,new Date(2016, 12, 16, 8, 41, 3, 409), 'AMS'),
	new GateCheckEvent(true,  new Date(2016, 12, 16, 17, 44, 56, 122), 'AMS'),
	new GateCheckEvent(false, new Date(2016, 12, 16, 18, 49, 4, 123), 'DH'),
	new GateCheckEvent(true,  new Date(2016, 12, 16, 22, 15, 44, 616), 'DH'),
	new GateCheckEvent(true,  new Date(2016, 12, 17, 8, 3, 54, 883), 'UTR'),
	new GateCheckEvent(false, new Date(2016, 12, 17, 8, 39, 21, 512), 'AMS')
));

exports.travelCostMatrix = getTravelCostMatrix();

exports.trainMetrics$ =  getTrainMetrics$();

function getTrainMetrics$() {
    simulationParameters = new TrainSimulationParameters(100, 140 / 3.6, 2 / 3.6, '1042', 40.0);

    simulation = new CompositeTrainSimulation(
        new StationaryTrainSimulation(RailwayStation.AMR.location, 60.0),
        new TrainJourneySimulation(RailwayStation.AMR.location, RailwayStation.UTR.location),
        new StationaryTrainSimulation(RailwayStation.UTR.location, 60.0)
    );
    
    return simulation.trainMetrics$(simulationParameters, Date.now());
}

exports.velocity$ = (trainMetrics$) => trainMetrics$
    .bufferCount(10, 5)
    .filter((measurement) => measurement.length > 1)
    .map(([first, ...rest]) => {
        const last = rest[rest.length-1];
        const elapsedTime = last.getTimestamp() - first.getTimestamp();
        const distance = last.getPosition().distanceTo(first.getPosition());
        const velocity = distance * 1000 / elapsedTime;
        return velocity;
    })
    .map((velocity) => velocity * 3.6);

function gateCheckEvent$() {
    if (gce == null) {
        gce = Rx.Observable.merge(
            singleGateCheckEvent$(true, 233), 
            singleGateCheckEvent$(true, 978),
            singleGateCheckEvent$(false,  1313),
		    singleGateCheckEvent$(true,   2105),
			singleGateCheckEvent$(false,  3643),
			singleGateCheckEvent$(false,  4411),
			singleGateCheckEvent$(true,   5556),
			singleGateCheckEvent$(false,  8123),
			singleGateCheckEvent$(false,  9722),
			singleGateCheckEvent$(true,  10880)
        );
    }
    return gce;
}

function singleGateCheckEvent$(isCheckin, delay) {
    return Rx.Observable.of(new GateCheckEvent(isCheckin, Date.now() + delay, 'AMR'))
        .delay(delay);
}

function getTravelCostMatrix() {
    const travelCosts = new TravelCostMatrix();
    travelCosts.addCostEntry('UTR', 'AMS', 7.50);
    travelCosts.addCostEntry('UTR', 'DH', 11.00);
    travelCosts.addCostEntry('DH', 'AMS', 11.50);
    return travelCosts;
}
